pkg/job_runner/awx_job_runner/awx_job_runner.go

```diff
@@ -14,34 +14,57 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package main
+package awx_job_runner
 
 import (
 	"fmt"
 
 	"github.com/golang/glog"
+	"golang.org/x/sync/syncmap"
+	"k8s.io/apimachinery/pkg/util/wait"
 
 	"github.com/moolitayer/awx-client-go/awx"
 	"github.com/openshift/autoheal/pkg/alertmanager"
 	"github.com/openshift/autoheal/pkg/apis/autoheal"
+
+	"github.com/openshift/autoheal/pkg/config"
+	"github.com/openshift/autoheal/pkg/metrics_exporter"
 )
 
-func (h *Healer) runAWXJob(rule *autoheal.HealingRule, action *autoheal.AWXJobAction, alert *alertmanager.Alert) error {
-	var err error
+type AWXJobRunner struct {
+	config *config.AWXConfig
+
+	activeJobs *syncmap.Map
+}
+
+func NewAWXJobRunner() *AWXJobRunner {
+	awxJubRunner := new(AWXJobRunner)
+	return awxJubRunner
+}
 
+func (r *AWXJobRunner) Init(config *config.Config, stopCh <-chan struct{}) *AWXJobRunner {
```

Comment:
This kind of constructor in two steps is problematic, in my opinion, specially the fact that the `Init` method receives multiple parameters. What if we need or want to add a new optional parameter in the future? We will have to change all the points where this method is called. That isn't very relevant in this case, because it is used only internally, but in general I prefer to use a builder:

```go
type Builder {
  ...
}

func NewBuilder() *Builder {
  return new(Builder)
}

// Config sets the configuration that will be used by the
// runner.
func (b *Builder) Config(config *config.Config) *Builder {
  b.config = config
  return b
}

// StopChannel sets the channel where the runner will wait
// for the signal to stop its background processing.
func (b *Builder) StopChannel(ch <-chan struct{}) *Builder {
  b.stopCh = ch
  return b
}

// Build creates a new runner using the details stored in
// the builder.
func (b *Builder) Build() (runner *AWXJobRunner, err error) {
  // Create and populate the runner, and also check for
  // possible errors, like missing mandatory parameters,
  // etc.
  runner = ...

  return
}
```

This way the user of this package can create the runner like this:

```go
runner, err := awxrunner.NewBuilder().
  Config(config).
  StopChannel(stopCh).
  Build()
if err != nil {
  ...
}
```

If we eventually decide to add optional parameters to that constructor, that code doesn't need to change.

Can you use this mechanism for both kinds of runners?

Reply:
Yeah very nice! Actoually I never really agreed with this approach in the code, but now i'm sold :+1: :trophy: